### PR TITLE
Add support for reduction in Yaksa

### PR DESCRIPTION
**Pull request by nusislam**
## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Have read and agree to the Yaksa CLA terms (https://github.com/pmodels/yaksa/wiki/Yaksa-Contributor-License-Agreement)

### CODE REVIEW

File: src/backend/gencomm.py [hunk 1]
```diff
@@ -11,8 +11,16 @@
...
+type_ops = {'_Bool': {'REPLACE', 'LAND', 'LOR', 'LXOR'},
+            'char': {'REPLACE'},
+            'wchar_t': {'REPLACE'},
+            'int8_t': {'REPLACE', 'SUM', 'PROD', 'MIN', 'MAX', 'LAND', 'LOR', 'LXOR', 'BAND', 'BOR', 'BXOR'},
+            'int16_t': {'REPLACE', 'SUM', 'PROD', 'MIN', 'MAX', 'LAND', 'LOR', 'LXOR', 'BAND', 'BOR', 'BXOR'},
+            'int32_t': {'REPLACE', 'SUM', 'PROD', 'MIN', 'MAX', 'LAND', 'LOR', 'LXOR', 'BAND', 'BOR', 'BXOR'},
+            'int64_t': {'REPLACE', 'SUM', 'PROD', 'MIN', 'MAX', 'LAND', 'LOR', 'LXOR', 'BAND', 'BOR', 'BXOR'},
+            'float': {'REPLACE', 'SUM', 'PROD', 'MIN', 'MAX'},
+            'double': {'REPLACE', 'SUM', 'PROD', 'MIN', 'MAX'},
+            'long double': {'REPLACE', 'SUM', 'PROD', 'MIN', 'MAX'}}
```
Comment (tarudoodi) [inline src/backend/gencomm.py:23]: Why is `byte` not listed here? Just curious since it is a basic datatype.

Comment (pavanbalaji) [reply src/backend/gencomm.py:23]: Not all types are passed to the backend.  For example, you'll notice that plain `int` is missing too.  It's converted to fewer types by the frontend before being passed to the backend.

Comment (tarudoodi) [reply src/backend/gencomm.py:23]: Ok, yes. Thank you.

File: src/backend/cuda/include/yaksuri_cudai.h [hunk 2]
```diff
@@ -79,10 +79,10 @@ int yaksuri_cudai_md_alloc(yaksi_type_s * type);
 int yaksuri_cudai_populate_pupfns(yaksi_type_s * type);
 
 int yaksuri_cudai_ipack(const void *inbuf, void *outbuf, uintptr_t count, yaksi_type_s * type,
-                        yaksi_info_s * info, int target);
+                        yaksi_info_s * info, yaksa_op_t op, int target);
 int yaksuri_cudai_iunpack(const void *inbuf, void *outbuf, uintptr_t count, yaksi_type_s * type,
-                          yaksi_info_s * info, int target);
-int yaksuri_cudai_pup_is_supported(yaksi_type_s * type, bool * is_supported);
+                          yaksi_info_s * info, yaksa_op_t op, int target);
+int yaksuri_cudai_pup_is_supported(yaksi_type_s * type, yaksa_op_t op, bool * is_supported);
```
Comment (nusislam) [inline src/backend/cuda/include/yaksuri_cudai.h:85]: The commit message seems a bit misleading. This commit has a lot of changes in the backend also.

Comment (pavanbalaji) [reply src/backend/cuda/include/yaksuri_cudai.h:85]: Yeah, the "frontend" tag needs to be removed.

Comment (pavanbalaji) [reply src/backend/cuda/include/yaksuri_cudai.h:85]: Other than that the message itself is fine, right?

Comment (nusislam) [reply src/backend/cuda/include/yaksuri_cudai.h:85]: Yes, if the tag is corrected, the rest seems fine.

Comment (pavanbalaji) [reply src/backend/cuda/include/yaksuri_cudai.h:85]: Fixed.

**Review by pavanbalaji — APPROVED** [review]

**Review by tarudoodi — APPROVED** [review]

**Pull request merged**